joint_trajectory_controller/src/joint_trajectory_controller.cpp

```diff
@@ -258,14 +337,153 @@ JointTrajectoryController::on_configure(const rclcpp_lifecycle::State &)
+      return CallbackReturn::ERROR;
+    }
+  }
+
+  // Check if command interfaces combination is valid. Valid combinations are:
+  // 1. effort
+  // 2. velocity
+  // 2. position [velocity, [acceleration]]
+
+  // effort can't be combined with other interfaces
+  if (check_if_interface_type_exists(command_interface_types_, hardware_interface::HW_IF_EFFORT)) {
+    if (command_interface_types_.size() == 1) {
+      // TODO(anyone): This flag is not used for now
+      // There should be PID-approach used as in ROS1:
+      // https://github.com/ros-controls/ros_controllers/blob/noetic-devel/joint_trajectory_controller/include/joint_trajectory_controller/hardware_interface_adapter.h#L283
+      use_closed_loop_pid_adapter = true;
+      // TODO(anyone): remove the next two lines when implemented
+      RCLCPP_ERROR(logger, "using 'effort' command interface alone is not yet implemented yet.");
+      return CallbackReturn::ERROR;
+    } else {
+      RCLCPP_ERROR(logger, "'effort' command interface has to be used alone.");
+      return CallbackReturn::ERROR;
+    }
+  }
+
+  if (check_if_interface_type_exists(
+      command_interface_types_, hardware_interface::HW_IF_VELOCITY))
+  {
+    // if there is only velocity then use also PID adapter
+    if (command_interface_types_.size() == 1) {
+      use_closed_loop_pid_adapter = true;
+      // TODO(anyone): remove this when implemented
+      RCLCPP_ERROR(logger, "using 'velocity' command interface alone is not yet implemented yet.");
+      return CallbackReturn::ERROR;
+      // if velocity interface can be used without position if multiple defined
+    } else if (!check_if_interface_type_exists(
+        command_interface_types_, hardware_interface::HW_IF_POSITION))
+    {
+      RCLCPP_ERROR(
+        logger, "'velocity' command interface can be used either alone or 'position' "
+        "interface has to be present.");
+      return CallbackReturn::ERROR;
+    }
+    // invalid: acceleration is defined and no velocity
+  } else if (check_if_interface_type_exists(
+      command_interface_types_, hardware_interface::HW_IF_ACCELERATION))
+  {
+    RCLCPP_ERROR(
+      logger, "'acceleration' command interface can only be used if 'velocity' and "
+      "'position' interfaces are present");
+    return CallbackReturn::ERROR;
+  }
+
+  // Read always state interfaces from the parameter because they can be used
+  // independently from the controller's type.
+  // Specialized, child controllers should set its default value.
+  state_interface_types_ = node_->get_parameter("state_interfaces").as_string_array();
+
+  if (state_interface_types_.empty()) {
+    RCLCPP_ERROR(logger, "'state_interfaces' parameter is empty.");
+    return CallbackReturn::ERROR;
+  }
+
+  if (check_if_interface_type_exists(state_interface_types_, hardware_interface::HW_IF_EFFORT)) {
+    RCLCPP_ERROR(logger, "State interface type 'effort' not allowed!");
```

Comment:
Because it cannot be used as state interface. Only position, velocity and acceleration are allowed. There is no way to calculate trajectory from effort.